java/fury-core/src/main/java/org/apache/fury/serializer/StringSerializer.java

```diff
@@ -235,75 +214,106 @@ public String readBytesString(MemoryBuffer buffer) {
-          Math.max(((int) (smoothByteArrayLength * 0.9 + numElements * 0.1)), DEFAULT_BUFFER_SIZE);
-      if (smoothByteArrayLength <= DEFAULT_BUFFER_SIZE) {
-        this.byteArray = new byte[DEFAULT_BUFFER_SIZE];
-      }
+  @CodegenInvoke
+  public String readCompressedCharsString(MemoryBuffer buffer) {
+    long header = buffer.readVarUint36Small();
+    byte coder = (byte) (header & 0b11);
+    int numBytes = (int) (header >>> 2);
+    char[] chars;
+    if (coder == LATIN1) {
+      chars = readCharsLatin1(buffer, numBytes);
+    } else if (coder == UTF8) {
+      chars = readCharsUTF8(buffer, numBytes);
+    } else if (coder == UTF16) {
+      chars = readCharsUTF16(buffer, numBytes);
+    } else {
+      throw new RuntimeException("Unknown coder type " + coder);
     }
-    return byteArray;
+    return newCharsStringZeroCopy(chars);
   }
 
   // Invoked by fury JIT
   public void writeJavaString(MemoryBuffer buffer, String value) {
     if (STRING_VALUE_FIELD_IS_BYTES) {
-      writeBytesString(buffer, value);
+      if (compressString) {
+        writeCompressedBytesString(buffer, value);
+      } else {
+        writeBytesString(buffer, value);
+      }
     } else {
       assert STRING_VALUE_FIELD_IS_CHARS;
-      final char[] chars = (char[]) Platform.getObject(value, STRING_VALUE_FIELD_OFFSET);
       if (compressString) {
-        if (StringUtils.isLatin(chars)) {
-          writeCharsLatin(buffer, chars, chars.length);
-        } else {
-          writeCharsUTF16(buffer, chars, chars.length);
-        }
+        writeCompressedCharsString(buffer, value);
       } else {
-        int numBytes = MathUtils.doubleExact(value.length());
-        buffer.writePrimitiveArrayWithSize(chars, Platform.CHAR_ARRAY_OFFSET, numBytes);
+        writeCharsString(buffer, value);
       }
     }
   }
 
+  public void writeUTF8String(MemoryBuffer buffer, String value) {
+    byte[] bytes = value.getBytes(StandardCharsets.UTF_8);
+    buffer.writeVarUint32(bytes.length);
+    buffer.writeBytes(bytes);
+  }
+
   // Invoked by fury JIT
   public String readJavaString(MemoryBuffer buffer) {
     if (STRING_VALUE_FIELD_IS_BYTES) {
-      return readBytesString(buffer);
+      if (compressString) {
+        return readCompressedBytesString(buffer);
+      } else {
+        return readBytesString(buffer);
+      }
     } else {
       assert STRING_VALUE_FIELD_IS_CHARS;
       if (compressString) {
         return readCompressedCharsString(buffer);
       } else {
-        return newCharsStringZeroCopy(buffer.readCharsAndSize());
+        return readCharsString(buffer);
       }
     }
   }
 
+  public void writeCompressedBytesString(MemoryBuffer buffer, String value) {
+    final byte[] bytes = (byte[]) Platform.getObject(value, STRING_VALUE_FIELD_OFFSET);
+    final byte coder = Platform.getByte(value, Offset.STRING_CODER_FIELD_OFFSET);
+    if (coder == LATIN1 || bestCoder(bytes) == UTF16) {
+      writeBytesString(buffer, value);
```

Comment:
Maybe we should make `writeBytesString` receive bytes and coder, and invoke like `writeBytesString(buffer, coder, bytes)`